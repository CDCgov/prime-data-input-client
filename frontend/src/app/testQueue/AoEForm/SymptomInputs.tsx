import React from "react";

import { globalSymptomDefinitions } from "../../../patientApp/timeOfTest/constants";

import Checkboxes from "../../commonComponents/Checkboxes";
import TextInput from "../../commonComponents/TextInput";

interface Symptoms {
  [key: string]: boolean;
}

interface Props {
  noSymptoms: boolean;
  setNoSymptoms: (noSymptoms: boolean) => void;
  currentSymptoms: Symptoms;
  setSymptoms: (symptoms: Symptoms) => void;
  onsetDate: string | undefined;
  setOnsetDate: (onsetDate: string) => void;
  symptomError: string | undefined;
  symptomOnsetError: string | undefined;
  symptomRef: React.RefObject<HTMLInputElement>;
  symptomOnsetRef: React.RefObject<HTMLInputElement>;
}

const SymptomInputs: React.FC<Props> = ({
  noSymptoms,
  setNoSymptoms,
  currentSymptoms,
  setSymptoms,
  onsetDate,
  setOnsetDate,
  symptomError,
  symptomOnsetError,
  symptomRef,
  symptomOnsetRef,
}) => {
  return (
    <>
      <div className={symptomError ? "usa-form-group--error" : "usa-fieldset"}>
        <Checkboxes
          name="no_symptoms"
          legend="Are you experiencing any of the following symptoms?"
          onChange={(e) => setNoSymptoms(e.target.checked)}
          boxes={[{ value: "no", label: "No Symptoms", checked: noSymptoms }]}
          required
          errorMessage={symptomError}
          validationStatus={symptomError ? "error" : undefined}
          inputRef={symptomRef}
        />
        {!noSymptoms && (
          <>
            <div className="border-top-1px border-base-lighter margin-y-2"></div>
            <Checkboxes
              legend="What are your symptoms?"
              legendSrOnly
              name="symptom_list"
              className="symptom-checkboxes"
              onChange={(e) =>
                setSymptoms({
                  ...currentSymptoms,
                  [e.target.value]: e.target.checked,
                })
              }
              boxes={globalSymptomDefinitions.map(({ label, value }) => ({
                label,
                value,
                checked: currentSymptoms[value],
              }))}
            />
          </>
        )}
      </div>
      {!noSymptoms && (
        <TextInput
          type="date"
          label="Date of symptom onset"
          name="symptom_onset"
          value={onsetDate}
          onChange={(e) => setOnsetDate(e.target.value)}
          min="2020-02-01"
          max={new Date().toISOString().split("T")[0]}
          required={Object.keys(currentSymptoms).some(
            (key) => currentSymptoms[key]
          )}
          errorMessage={symptomOnsetError}
          validationStatus={symptomOnsetError ? "error" : undefined}
          inputRef={symptomOnsetRef}
        />
      )}
    </>
  );
};

export default SymptomInputs;
