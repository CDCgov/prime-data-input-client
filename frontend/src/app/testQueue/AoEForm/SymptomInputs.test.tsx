import renderer from 'react-test-renderer';
import MockDate from 'mockdate';

import SymptomInputs from './SymptomInputs';

describe('SymptomInputs', () => {
  let component: renderer.ReactTestRenderer;
  let setNoSymptoms: jest.Mock;
  beforeEach(() => {
    MockDate.set('2021-02-08');
    setNoSymptoms = jest.fn();
    component = renderer.create(
      <SymptomInputs
        noSymptoms={false}
        setNoSymptoms={setNoSymptoms}
        currentSymptoms={{}}
        setSymptoms={jest.fn()}
        onsetDate={''}
        setOnsetDate={jest.fn()}
        symptomError={undefined}
        symptomOnsetError={undefined}
      />
    );
  });

  it('renders', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('setting has symptoms', () => {
    describe('no symptoms', () => {
      beforeEach(async () => {
        renderer.act(() => {
          component.root
            .findByProps({ name: 'no_symptoms' })
            .props.onChange({ target: { checked: false } });
        });
      });
      it('calls setNoSymptoms', () => {
        expect(setNoSymptoms).toHaveBeenCalledTimes(1);
      });
      it('calls setNoSymptoms with true', () => {
        expect(setNoSymptoms).toHaveBeenCalledWith(false);
      });
    });
    describe('symptoms', () => {
      beforeEach(async () => {
        renderer.act(() => {
          component.root
            .findByProps({ name: 'no_symptoms' })
            .props.onChange({ target: { checked: true } });
        });
      });
      it('calls setNoSymptoms', () => {
        expect(setNoSymptoms).toHaveBeenCalledTimes(1);
      });
      it('calls setNoSymptoms with true', () => {
        expect(setNoSymptoms).toHaveBeenCalledWith(true);
      });
    });
  });
});
