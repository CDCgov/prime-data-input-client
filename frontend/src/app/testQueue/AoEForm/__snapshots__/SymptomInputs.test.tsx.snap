// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SymptomInputs renders 1`] = `
Array [
  <div
    className="usa-fieldset"
  >
    <fieldset
      className="usa-fieldset prime-checkboxes"
    >
      <legend
        className="usa-legend"
      >
        Are you experiencing any of the following symptoms?
        <abbr
          className="usa-hint usa-hint--required"
          title="required"
        >
           
          *
        </abbr>
      </legend>
      <div
        className="usa-form-group"
      >
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="1-val-0"
            name="no_symptoms"
            onChange={[Function]}
            type="checkbox"
            value="no"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="1-val-0"
          >
            No Symptoms
          </label>
        </div>
      </div>
    </fieldset>
    <div
      className="border-top-1px border-base-lighter margin-y-2"
    />
    <fieldset
      className="usa-fieldset prime-checkboxes symptom-checkboxes"
    >
      <legend
        className="usa-legend usa-sr-only"
      >
        What are your symptoms?
      </legend>
      <div
        className="usa-form-group"
      >
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-0"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="426000000"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-0"
          >
            Fever over 100.4F
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-1"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="103001002"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-1"
          >
            Feeling feverish
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-2"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="43724002"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-2"
          >
            Chills
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-3"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="49727002"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-3"
          >
            Cough
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-4"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="267036007"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-4"
          >
            Shortness of breath
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-5"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="230145002"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-5"
          >
            Difficulty breathing
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-6"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="84229001"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-6"
          >
            Fatigue
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-7"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="68962001"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-7"
          >
            Muscle or body aches
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-8"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="25064002"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-8"
          >
            Headache
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-9"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="36955009"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-9"
          >
            New loss of taste
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-10"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="44169009"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-10"
          >
            New loss of smell
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-11"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="162397003"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-11"
          >
            Sore throat
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-12"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="68235000"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-12"
          >
            Nasal congestion
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-13"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="64531003"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-13"
          >
            Runny nose
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-14"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="422587007"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-14"
          >
            Nausea
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-15"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="422400008"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-15"
          >
            Vomiting
          </label>
        </div>
        <div
          className="usa-checkbox"
        >
          <input
            checked={false}
            className="usa-checkbox__input"
            id="2-val-16"
            name="symptom_list"
            onChange={[Function]}
            type="checkbox"
            value="62315008"
          />
          <label
            className="usa-checkbox__label"
            htmlFor="2-val-16"
          >
            Diarrhea
          </label>
        </div>
      </div>
    </fieldset>
  </div>,
  <div
    className="usa-form-group"
  >
    <label
      className="usa-label"
      htmlFor="3"
    >
      Date of symptom onset
    </label>
    <input
      aria-required="false"
      className="usa-input"
      id="3"
      max="2021-02-08"
      min="2020-02-01"
      name="symptom_onset"
      onChange={[Function]}
      type="date"
      value=""
    />
  </div>,
]
`;
